cachi2/core/package_managers/pip.py

```diff
@@ -1411,12 +1423,26 @@ def _download_dependencies(
 
         if require_hashes or req.kind == "url":
             hashes = req.hashes or [req.qualifiers["cachito_hash"]]
-            _verify_hash(download_info["path"], hashes)
+            _verify_hash(download_info["path"], list(map(_to_checksum_info, hashes)))
 
         download_info["kind"] = req.kind
-        downloads.append(download_info)
+        downloaded.append(download_info)
+
+    if allow_binary:
+        log.info("Downloading %d wheel(s) ...", len(to_download))
+        files: dict[str, Union[str, PathLike[str]]] = {pkg.url: pkg.path for pkg in to_download}
+        asyncio.run(async_download_files(files, get_config().concurrency_limit))
+
+        for pkg in to_download:
+            try:
+                _verify_hash(pkg.path, pkg.checksums)
+            except PackageRejected:
+                pkg.path.unlink()
+                log.warning("The file is removed from the output directory")
```

Comment:
One thing that IMO makes this confusing is that `pkg.checksums` can have two different meanings here

a) the hashes for a single wheel file as reported by PyPI
b) all the `--hash` options for a package as specified by the user

How about we use `pkg.checksums` only for checksums reported by PyPI? And when we want to verify the user-specified checksums, we would take them from `user_checksums` instead? Would that be clearer?

Maybe it would also simplify the `_process_package_distributions` function, which is becoming quite complex